// @ngxs/store/esm2015/ngxs-store.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { NgxsModule, Action, Store, State, Select, SelectorOptions, Actions, getSelectorMetadata, getStoreMetadata, ensureStoreMetadata, ensureSelectorMetadata, ofAction, ofActionDispatched, ofActionSuccessful, ofActionCanceled, ofActionErrored, ofActionCompleted, NgxsSimpleChange, Selector, getActionTypeFromInstance, actionMatcher, createSelector, NoopNgxsExecutionStrategy, StateToken, NGXS_PLUGINS, StateStream, setValue, getValue, InitState, UpdateState } from './index';
export { InternalActions as ɵb, OrderedSubject as ɵa } from './src/actions-stream';
export { SelectFactory as ɵs } from './src/decorators/select/select-factory';
export { InternalNgxsExecutionStrategy as ɵq } from './src/execution/internal-ngxs-execution-strategy';
export { NGXS_EXECUTION_STRATEGY as ɵi, USER_PROVIDED_NGXS_EXECUTION_STRATEGY as ɵh } from './src/execution/symbols';
export { InternalDispatchedActionResults as ɵl, InternalDispatcher as ɵm } from './src/internal/dispatcher';
export { InternalErrorReporter as ɵr } from './src/internal/error-handler';
export { ensureSelectorMetadata as ɵw, ensureStoreMetadata as ɵu, getSelectorMetadata as ɵx, getStoreMetadata as ɵv } from './src/internal/internals';
export { LifecycleStateManager as ɵy } from './src/internal/lifecycle-state-manager';
export { StateContextFactory as ɵn } from './src/internal/state-context-factory';
export { StateFactory as ɵk } from './src/internal/state-factory';
export { InternalStateOperations as ɵo } from './src/internal/state-operations';
export { NgxsFeatureModule as ɵz } from './src/modules/ngxs-feature.module';
export { NgxsRootModule as ɵj } from './src/modules/ngxs-root.module';
export { PluginManager as ɵp } from './src/plugin-manager';
export { FEATURE_STATE_TOKEN as ɵd, NgxsConfig as ɵf, ROOT_STATE_TOKEN as ɵc, SELECTOR_META_KEY as ɵe } from './src/symbols';
export { mergeDeep as ɵg } from './src/utils/utils';
